ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/MyAppPurchaseLogic.kt

```diff
@@ -0,0 +1,118 @@
+package com.revenuecat.purchases.ui.revenuecatui
+
+import android.app.Activity
+import com.android.billingclient.api.Purchase
+import com.revenuecat.purchases.CustomerInfo
+import com.revenuecat.purchases.Package
+import com.revenuecat.purchases.PresentedOfferingContext
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.models.GoogleReplacementMode
+import com.revenuecat.purchases.models.PurchaseState
+import com.revenuecat.purchases.models.PurchaseType
+import com.revenuecat.purchases.models.StoreTransaction
+import org.json.JSONObject
+
+/**
+ * Code for handling in-app purchases and restorations directly by the application rather than by RevenueCat.
+ * These methods are called by a RevenueCat Paywall only when `Purchases.purchasesAreCompletedBy` is `MY_APP`.
+ *
+ * @property performPurchase Performs an in-app purchase.
+ * @property performRestore Perform a purchase restore for the given customer.
+ */
+class MyAppPurchaseLogic(
```

Comment:
It would be great if this implemented an interface `PurchaseLogic`. That way, we could have our own internal implementation of it (e.g. `RevenueCatPurchaseLogic`) which contains the logic that exists now (before this PR). 

That would make especially the `PaywallViewModel` a bit more straightforward imo. `performPurchase()` and `performRestore()` would simply check which `PurchaseLogic` to forward the calls to. 

Just an idea that _seems_ feasible, but let me know if it isn't!

Reply:
Yes I think this is a good idea, but will wait until I have direction on whether to co-routines or not.

Reply:
What do you think about renaming `MyAppPurchaseLogic` to just `PurchaseLogic`? 

That way we keep the option open to ourselves to put our own logic in an internal implementation of `PurchaseLogic`, e.g. called `RevenueCatPurchaseLogic`. I think doing this refactor is out of scope of this PR, but the suggested rename would leave that door open.

Reply:
I'm not sure, it was named `MyAppPurchaseLogic` in reference to PurchasesAreCompletedBy.MY_APP for which it is needed. If we need to implement something internally in the future, perhaps we can live with the "confusion" of having the base interface call "My App", because I think it does help customers connect the dots.

Reply:
Actually, I'll change it.

Reply:
Done here: https://github.com/RevenueCat/purchases-android/pull/1777/commits/d6683f77cfb6d3e135adf4fbe0cd85067deabee6

Also renamed the MyApp purchase/restore result types. Because `PurchaseResult` is already taken, it is `PurchaseLogicPurchaseResult` and `PurchaseLogicRestoreResult`.

Reply:
You do make a good point haha. It's not a blocker for me in any way.